### ISSUE TITLE
Simulator and helper code

### ISSUE THREAD

**Issue by gentlegiantJGC** [standupmaths/xmastree2021#1]
Hi, I am looking forward to this year's video.

Over the last couple of days I have written a simulator and some utility code to help make users code a bunch simpler than last year.

Are these things you would be interested including so that others can use them?

They will probably need modifying a little to work with the CSV format and if anything else has changed.

Simulator:
https://github.com/gentlegiantJGC/xmastree2020/blob/main/simulator.py

Utility code to import either the real tree or fall back to the simulator if it does not exist; a function to set up the LED interface that was imported and a context manager to control the frame time.
https://github.com/gentlegiantJGC/xmastree2020/blob/main/matts_tree_helpers.py

I also added a minimal template using this system
https://github.com/gentlegiantJGC/xmastree2020/blob/main/template.py

**Comment by gentlegiantJGC** [comment]
I sent this before realising the video had been released and have since watched the video and saw that you are only accepting spreadsheet submissions.

I may modify this code to take the gift files as input to visualise unless I find another visualiser. I couldn't see one in the other repository.

I may also look into adding a command line argument to the visualiser to make it also output as the gift format as the code runs so python users don't have to do their own file writing (again I didn't see anything for this but didn't look too hard)

It is also a little sad it can't randomise and it must be designed to loop or have a jump at some point when it restarts but I understand your reasoning.

**Comment by leukipp** [comment]
> I may modify this code to take the CSV animation files as input to visualise unless I find another visualiser. I couldn't see one in the other repository.

I have made a GIFT-Simulator that runs independent of the programming language used.

It`s available as online application here (still work in progress):
[https://leukipp.github.io/xmastree2021/simulator/](https://leukipp.github.io/xmastree2021/simulator/)

A procedural tree is rendered and light coordinates/animation frames can be loaded from url`s or local csv files.

**Comment by gentlegiantJGC** [comment]
> I have made a GIFT-Simulator 
That is cool and I like the addition of the tree.

I think technically the GIFT format was just the coordinate scaling which can be packed into a CSV file or the old text format linked in Matt's video description.

I don't think the animation format was actually given a name.

My plan is that this module can be used to simulate python code that could be on the real tree and also generate the CSV animation file as the program runs.

**Comment by leukipp** [comment]
> My plan is that this module can be used to simulate python code that could be on the real tree and also generate the CSV animation file as the program runs.

At the moment you can interact with the application via url parameters as described here:
[https://github.com/leukipp/xmastree2021/blob/main/simulator/README.md](https://github.com/leukipp/xmastree2021/blob/main/simulator/README.md)

It should also be possible to run everything as a standalone application, so that one has direct access to the file system (without the hassle of uploading csv files).

**Comment by oliverdunk** [comment]
With regards to simulators, my plan is to go through what's been PR'd and pick a simple one to merge to the repository. I think it'd be nice to have a trusted one that can be used for verifying submissions and that everyone can contribute to.

Beyond that, I think simulators are best in the "further work" section of the README so we're not filling the repo with them :)

TLDR: Feel free to PR something, but I'm probably only going to merge one :)
